File: onyx-database-tests/test/remote/queries/QueryListenerTest.java
package remote.queries;

import category.RemoteServerTests;
import com.onyx.exception.OnyxException;
import com.onyx.persistence.IManagedEntity;
import com.onyx.persistence.query.Query;
import com.onyx.persistence.query.QueryListener;
import entities.SimpleEntity;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import remote.base.RemoteBaseTest;

import java.io.IOException;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Created by tosborn1 on 3/27/17.
 *
 * This unit test covers registered query listeners and the events
 */
@Category(value = RemoteServerTests.class)
public class QueryListenerTest extends RemoteBaseTest {

    @After
    public void after() throws IOException {
        shutdown();
    }

    @Before
    public void before() throws OnyxException {
        initialize();
        seedData();
    }

    public void seedData() throws OnyxException
    {
        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("1");
        simpleEntity.setName("1");
        manager.saveEntity(simpleEntity);
    }

    /**
     * Test a query is subscribed correctly
     */
    @Test
    public void testSubscribe() throws OnyxException
    {
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity items) {

            }

            @Override
            public void onItemAdded(IManagedEntity items) {

            }

            @Override
            public void onItemRemoved(IManagedEntity items) {

            }
        });

        manager.executeQuery(query);

    }

    /**
     * Test a query is un-subscribed correctly
     */
    @Test
    public void testUnsubscribe() throws OnyxException, InterruptedException {
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity items) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity items) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity items) {
                assert false;
            }
        });

        manager.executeQuery(query);

        manager.removeChangeListener(query);


        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("23");
        simpleEntity.setName("23");
        manager.saveEntity(simpleEntity);

        // Ensure it is not duplicating listeners
        manager.executeQuery(query);

        Thread.sleep(1000);


    }

    /**
     * Test an entity correctly hits an insert listener
     */
    @Test
    public void testInsert() throws OnyxException, InterruptedException {
        final AtomicBoolean pass = new AtomicBoolean(false);
        final CountDownLatch countDownLatch = new CountDownLatch(1);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("22");
                assert ((SimpleEntity) entity).getName().equals("22");
                countDownLatch.countDown();
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                assert false;
            }
        });

        manager.executeQuery(query);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("22");
        simpleEntity.setName("22");
        manager.saveEntity(simpleEntity);

        countDownLatch.await(5, TimeUnit.SECONDS);

        assert pass.get();
    }

    /**
     * Test an entity correctly hits an update listener
     */
    @Test
    public void testUpdate() throws OnyxException, InterruptedException {
        final CountDownLatch countDownLatch = new CountDownLatch(1);

        final AtomicBoolean pass = new AtomicBoolean(false);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("1");
                assert ((SimpleEntity) entity).getName().equals("2");
                countDownLatch.countDown();
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                assert false;
            }
        });

        manager.executeQuery(query);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("1");
        simpleEntity.setName("2");
        manager.saveEntity(simpleEntity);

        countDownLatch.await(5, TimeUnit.SECONDS);

        assert pass.get();
    }

    /**
     * Test an entity correctly hits an delete listener
     */
    @Test
    public void testDelete() throws OnyxException, InterruptedException {
        final CountDownLatch countDownLatch = new CountDownLatch(1);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("55");
        simpleEntity.setName("2");
        manager.saveEntity(simpleEntity);

        final AtomicBoolean pass = new AtomicBoolean(false);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("55");
                assert ((SimpleEntity) entity).getName().equals("2");
                countDownLatch.countDown();
            }
        });

        manager.executeQuery(query);

        simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("55");
        simpleEntity.setName("2");
        manager.deleteEntity(simpleEntity);

        countDownLatch.await(5, TimeUnit.SECONDS);

        assert pass.get();
    }


    /**
     * Test a query is subscribed correctly
     */
    @Test
    public void testSubscribeListen() throws OnyxException
    {
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity items) {

            }

            @Override
            public void onItemAdded(IManagedEntity items) {

            }

            @Override
            public void onItemRemoved(IManagedEntity items) {

            }
        });

        manager.listen(query);

    }

    /**
     * Test a query is un-subscribed correctly
     */
    @Test
    public void testUnsubscribeListen() throws OnyxException, InterruptedException {
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity items) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity items) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity items) {
                assert false;
            }
        });

        manager.listen(query);

        manager.removeChangeListener(query);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("23");
        simpleEntity.setName("23");
        manager.saveEntity(simpleEntity);


        Thread.sleep(1000);

    }

    /**
     * Test an entity correctly hits an insert listener
     */
    @Test
    public void testInsertListen() throws OnyxException, InterruptedException {
        final CountDownLatch countDownLatch = new CountDownLatch(1);

        final AtomicBoolean pass = new AtomicBoolean(false);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("220");
                assert ((SimpleEntity) entity).getName().equals("22");
                countDownLatch.countDown();
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                assert false;
            }
        });

        manager.listen(query);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("220");
        simpleEntity.setName("22");
        manager.saveEntity(simpleEntity);

        assert countDownLatch.await(1, TimeUnit.SECONDS);

        assert pass.get();

    }

    /**
     * Test an entity correctly hits an update listener
     */
    @Test
    public void testUpdateListen() throws OnyxException, InterruptedException {
        final CountDownLatch countDownLatch = new CountDownLatch(1);
        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("11");
        simpleEntity.setName("2");
        manager.saveEntity(simpleEntity);

        final AtomicBoolean pass = new AtomicBoolean(false);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("11");
                assert ((SimpleEntity) entity).getName().equals("2");
                countDownLatch.countDown();
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                assert false;
            }
        });

        manager.listen(query);

        simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("11");
        simpleEntity.setName("2");
        manager.saveEntity(simpleEntity);

        assert countDownLatch.await(1, TimeUnit.SECONDS);
        assert pass.get();
    }

    /**
     * Test an entity correctly hits an delete listener
     */
    @Test
    public void testDeleteListen() throws OnyxException, InterruptedException {
        final CountDownLatch countDownLatch = new CountDownLatch(1);

        SimpleEntity simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("12");
        simpleEntity.setName("2");
        manager.saveEntity(simpleEntity);

        final AtomicBoolean pass = new AtomicBoolean(false);
        Query query = new Query(SimpleEntity.class);
        query.setChangeListener(new QueryListener<IManagedEntity>() {

            @Override
            public void onItemUpdated(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemAdded(IManagedEntity entity) {
                assert false;
            }

            @Override
            public void onItemRemoved(IManagedEntity entity) {
                pass.set(true);
                assert entity instanceof SimpleEntity;
                assert ((SimpleEntity) entity).getSimpleId().equals("12");
                assert ((SimpleEntity) entity).getName().equals("2");
                countDownLatch.countDown();
            }
        });

        manager.listen(query);

        simpleEntity = new SimpleEntity();
        simpleEntity.setSimpleId("12");
        simpleEntity.setName("2");
        manager.deleteEntity(simpleEntity);

        assert countDownLatch.await(1, TimeUnit.SECONDS);

        assert pass.get();
    }
}

File: onyx-database-tests/test/remote/queries/QueryListenerTest.kt
package remote.queries

import category.RemoteServerTests
import com.onyx.application.impl.DatabaseServer
import com.onyx.exception.OnyxException
import com.onyx.persistence.IManagedEntity
import com.onyx.persistence.query.Query
import com.onyx.persistence.query.QueryListener
import entities.SimpleEntity
import org.junit.After
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.experimental.categories.Category
import remote.base.RemoteBaseTest

import java.io.IOException
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Created by Tim Osborn on 3/27/17.
 *
 * This unit test covers registered query listeners and the events
 */
@Category(value = RemoteServerTests::class)
class QueryListenerTest : RemoteBaseTest() {

    @After
    @Throws(IOException::class)
    fun after() {
        shutdown()
    }

    @Before
    @Throws(OnyxException::class)
    fun before() {
        initialize()
        seedData()
    }

    @Throws(OnyxException::class)
    fun seedData() {
        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "1"
        simpleEntity.name = "1"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)
    }

    /**
     * Test a query is subscribed correctly
     */
    @Test
    @Throws(OnyxException::class)
    fun testSubscribe() {
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(items: IManagedEntity) {

            }

            override fun onItemAdded(items: IManagedEntity) {

            }

            override fun onItemRemoved(items: IManagedEntity) {

            }
        }

        manager!!.executeQuery<Any>(query)

    }

    /**
     * Test a query is un-subscribed correctly
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testUnsubscribe() {
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(items: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(items: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(items: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.executeQuery<Any>(query)

        manager!!.removeChangeListener(query)


        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "23"
        simpleEntity.name = "23"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        // Ensure it is not duplicating listeners
        manager!!.executeQuery<Any>(query)

        Thread.sleep(1000)


    }

    /**
     * Test an entity correctly hits an insert listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testInsert() {
        val pass = AtomicBoolean(false)
        val countDownLatch = CountDownLatch(1)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "22")
                assert(entity.name == "22")
                countDownLatch.countDown()
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.executeQuery<Any>(query)

        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "22"
        simpleEntity.name = "22"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        countDownLatch.await(5, TimeUnit.SECONDS)

        assert(pass.get())
    }

    /**
     * Test an entity correctly hits an update listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testUpdate() {
        val countDownLatch = CountDownLatch(1)

        val pass = AtomicBoolean(false)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "1")
                assert(entity.name == "2")
                countDownLatch.countDown()
            }

            override fun onItemAdded(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.executeQuery<Any>(query)

        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "1"
        simpleEntity.name = "2"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        countDownLatch.await(5, TimeUnit.SECONDS)

        assert(pass.get())
    }

    /**
     * Test an entity correctly hits an delete listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testDelete() {
        val countDownLatch = CountDownLatch(1)

        var simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "55"
        simpleEntity.name = "2"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        val pass = AtomicBoolean(false)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "55")
                assert(entity.name == "2")
                countDownLatch.countDown()
            }
        }

        manager!!.executeQuery<Any>(query)

        simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "55"
        simpleEntity.name = "2"
        manager!!.deleteEntity(simpleEntity)

        countDownLatch.await(5, TimeUnit.SECONDS)

        assert(pass.get())
    }


    /**
     * Test a query is subscribed correctly
     */
    @Test
    @Throws(OnyxException::class)
    fun testSubscribeListen() {
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(items: IManagedEntity) {

            }

            override fun onItemAdded(items: IManagedEntity) {

            }

            override fun onItemRemoved(items: IManagedEntity) {

            }
        }

        manager!!.listen(query)

    }

    /**
     * Test a query is un-subscribed correctly
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testUnsubscribeListen() {
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(items: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(items: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(items: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.listen(query)

        manager!!.removeChangeListener(query)

        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "23"
        simpleEntity.name = "23"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)


        Thread.sleep(1000)

    }

    /**
     * Test an entity correctly hits an insert listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testInsertListen() {
        val countDownLatch = CountDownLatch(1)

        val pass = AtomicBoolean(false)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "220")
                assert(entity.name == "22")
                countDownLatch.countDown()
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.listen(query)

        val simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "220"
        simpleEntity.name = "22"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        assert(countDownLatch.await(1, TimeUnit.SECONDS))

        assert(pass.get())

    }

    /**
     * Test an entity correctly hits an update listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testUpdateListen() {
        val countDownLatch = CountDownLatch(1)
        var simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "11"
        simpleEntity.name = "2"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        val pass = AtomicBoolean(false)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "11")
                assert(entity.name == "2")
                countDownLatch.countDown()
            }

            override fun onItemAdded(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                assert(false)
            }
        }

        manager!!.listen(query)

        simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "11"
        simpleEntity.name = "2"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        assert(countDownLatch.await(1, TimeUnit.SECONDS))
        assert(pass.get())
    }

    /**
     * Test an entity correctly hits an delete listener
     */
    @Test
    @Throws(OnyxException::class, InterruptedException::class)
    fun testDeleteListen() {
        val countDownLatch = CountDownLatch(1)

        var simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "12"
        simpleEntity.name = "2"
        manager!!.saveEntity<IManagedEntity>(simpleEntity)

        val pass = AtomicBoolean(false)
        val query = Query(SimpleEntity::class.java)
        query.changeListener = object : QueryListener<IManagedEntity> {

            override fun onItemUpdated(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemAdded(entity: IManagedEntity) {
                assert(false)
            }

            override fun onItemRemoved(entity: IManagedEntity) {
                pass.set(true)
                assert(entity is SimpleEntity)
                assert((entity as SimpleEntity).simpleId == "12")
                assert(entity.name == "2")
                countDownLatch.countDown()
            }
        }

        manager!!.listen(query)

        simpleEntity = SimpleEntity()
        simpleEntity.simpleId = "12"
        simpleEntity.name = "2"
        manager!!.deleteEntity(simpleEntity)

        assert(countDownLatch.await(1, TimeUnit.SECONDS))

        assert(pass.get())
    }
}

